### PR TITLE
Modernize cunit recipe

### PR DESCRIPTION
**Pull request by Pro**
Specify library name and version:  **cunit/2.1-3**

Modernize cunit recipe for better Conan 2 support and fixing build on windows

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/cunit/all/conanfile.py [hunk 1]
```diff
@@ -57,89 +63,79 @@ def _settings_build(self):
...
 
     def build_requirements(self):
-        self.build_requires("libtool/2.4.6")
-        if self._settings_build.os == "Windows" and not tools.get_env("CONAN_BASH_PATH"):
-            self.build_requires("msys2/cci.latest")
+        self.build_requires("libtool/2.4.7")
+        if self._settings_build.os == "Windows":
+            self.win_bash = True
+            self.build_requires("autoconf/2.71")
+            self.tool_requires("automake/1.16.5")  # Needed for complie and lib wrappers
```
Comment (SpaceIm) [inline recipes/cunit/all/conanfile.py:70]: Never, it doesn't scale (it's what we were doing in the past and it was a nightmare of versions conflicts).

File: recipes/cunit/all/conanfile.py [hunk 2]
```diff
@@ -57,89 +63,79 @@ def _settings_build(self):
         return getattr(self, "settings_build", self.settings)
 
     def build_requirements(self):
-        self.build_requires("libtool/2.4.6")
-        if self._settings_build.os == "Windows" and not tools.get_env("CONAN_BASH_PATH"):
-            self.build_requires("msys2/cci.latest")
+        self.build_requires("libtool/2.4.7")
```
Comment (SpaceIm) [inline recipes/cunit/all/conanfile.py:66]:
```suggestion
        self.tool_requires("libtool/2.4.7")
```

File: recipes/cunit/all/conanfile.py [hunk 3]
```diff
@@ -57,89 +63,79 @@ def _settings_build(self):
...
-            conf_args.extend(["--disable-shared", "--enable-static"])
-        self._autotools.configure(args=conf_args, host=host, build=build)
-        return self._autotools
+    def generate(self):
+        env = VirtualBuildEnv(self)
+        env.generate()
+
+        tc = AutotoolsToolchain(self)
+        if (self.settings.get_safe("compiler") == "Visual Studio" and self.settings.get_safe("compiler.version") >= "12") or \
+                (self.settings.get_safe("compiler") == "msvc" and self.settings.get_safe("compiler.version") >= "180"):
```
Comment (SpaceIm) [inline recipes/cunit/all/conanfile.py:90]: prefer check_min_vs please

File: recipes/cunit/all/conanfile.py [hunk 4]
```diff
@@ -57,89 +63,79 @@ def _settings_build(self):
...
+            tc.extra_cflags.append("-FS")
+            tc.extra_cxxflags.append("-FS")
+        env = tc.environment()
```
Comment (SpaceIm) [inline recipes/cunit/all/conanfile.py:93]:
```suggestion
```

File: recipes/cunit/all/conanfile.py [hunk 5]
```diff
@@ -57,89 +63,79 @@ def _settings_build(self):
...
+        tc.configure_args.append("--datarootdir={}".format(os.path.join(self.package_folder, "bin", "share").replace("\\", "/")))
+        tc.configure_args.append("--enable-debug" if self.settings.build_type == "Debug" else "--disable-debug")
+        tc.configure_args.append("--enable-automated" if self.options.enable_automated else "--disable-automated")
+        tc.configure_args.append("--enable-basic" if self.options.enable_basic else "--disable-basic")
+        tc.configure_args.append("--enable-console" if self.options.enable_console else "--disable-console")
+        tc.configure_args.append("--enable-curses" if self.options.with_curses is not False else "--disable-curses")
+
```
Comment (SpaceIm) [inline recipes/cunit/all/conanfile.py:100]:
```suggestion
        env = tc.environment()
```

**Review by SpaceIm — CHANGES_REQUESTED** [review]

File: recipes/cunit/all/conanfile.py [hunk 6]
```diff
@@ -1,9 +1,16 @@
...
+from conan import ConanFile
+from conan.tools.microsoft import is_msvc, unix_path
+from conan.tools.files import apply_conandata_patches, copy, export_conandata_patches, get, rmdir
+from conan.tools.gnu import Autotools, AutotoolsToolchain
+from conan.tools.env import VirtualBuildEnv
 import glob
 import os
+from conan.tools.layout import basic_layout
+from conans.tools import chdir, remove_files_by_mask, rename
```
Comment (SpaceIm) [inline recipes/cunit/all/conanfile.py:11]: there are conan v2 flavors of these functions in conan.tools.files

File: recipes/cunit/all/conanfile.py [hunk 7]
```diff
@@ -30,23 +37,22 @@ class CunitConan(ConanFile):
...
+        export_conandata_patches(self)
 
-    @property
-    def _source_subfolder(self):
-        return "source_subfolder"
+    def configure(self):
+        if self.options.shared:
+            self.options.rm_safe("fPIC")
+        self.settings.rm_safe("compiler.libcxx")
+        self.settings.rm_safe("compiler.cppstd")
```
Comment (SpaceIm) [inline recipes/cunit/all/conanfile.py:48]: keep configure() after config_options() please

File: recipes/cunit/all/conanfile.py [hunk 8]
```diff
@@ -57,89 +63,79 @@ def _settings_build(self):
...
-        tools.get(**self.conan_data["sources"][self.version], strip_root=True, destination=self._source_subfolder)
-        with tools.chdir(self._source_subfolder):
+        get(self, **self.conan_data["sources"][self.version], strip_root=True)
+        with chdir(self.source_folder):
             for f in glob.glob("*.c"):
                 os.chmod(f, 0o644)
 
     @property
     def _user_info_build(self):
         return getattr(self, "user_info_build", self.deps_user_info)
```
Comment (SpaceIm) [inline recipes/cunit/all/conanfile.py:83]:
```suggestion
```

not needed anymore

File: recipes/cunit/all/conanfile.py [hunk 9]
```diff
@@ -57,89 +63,79 @@ def _settings_build(self):
...
+        env = VirtualBuildEnv(self)
+        env.generate()
+
+        tc = AutotoolsToolchain(self)
+        if (self.settings.get_safe("compiler") == "Visual Studio" and self.settings.get_safe("compiler.version") >= "12") or \
+                (self.settings.get_safe("compiler") == "msvc" and self.settings.get_safe("compiler.version") >= "180"):
+            tc.extra_cflags.append("-FS")
+            tc.extra_cxxflags.append("-FS")
+        env = tc.environment()
+        tc.configure_args.append("--datarootdir={}".format(os.path.join(self.package_folder, "bin", "share").replace("\\", "/")))
```
Comment (SpaceIm) [inline recipes/cunit/all/conanfile.py:94]:
```suggestion
        tc.configure_args.append("--datarootdir=${prefix}/bin/share")
```

File: recipes/cunit/all/conanfile.py (hunk 1 above)
Comment (Pro) [reply recipes/cunit/all/conanfile.py]: @SpaceIm I tried to remove the `automake` dependency, but then the build fails on windows.

Check the generate method, it uses `self.conf.get("user.automake:compile-wrapper")` and this one is None, if automake dependency is not given...

Comment (SpaceIm) [reply recipes/cunit/all/conanfile.py]: Yes, it seems that `self.conf.get()` can't see conf coming from transitive dependencies.
But you can use `self.dependencies.build["automake"].conf_info.get("user.automake:compile-wrapper", check_type=str)` instead.

File: recipes/cunit/all/conanfile.py [hunk 10]
```diff
@@ -30,116 +37,113 @@ class CunitConan(ConanFile):
...
-        else:
-            conf_args.extend(["--disable-shared", "--enable-static"])
-        self._autotools.configure(args=conf_args, host=host, build=build)
-        return self._autotools
+    def generate(self):
+        env = VirtualBuildEnv(self)
+        env.generate()
+
+        tc = AutotoolsToolchain(self)
+        if check_min_vs(self, "180", raise_invalid=False):
```
Comment (SpaceIm) [inline recipes/cunit/all/conanfile.py:95]:
```suggestion
        if is_msvc(self) and check_min_vs(self, "180", raise_invalid=False):
```

File: recipes/cunit/all/conanfile.py [hunk 11]
```diff
@@ -30,116 +37,113 @@ class CunitConan(ConanFile):
         "enable_console": True,
         "with_curses": False,
     }
-    exports_sources = "patches/**"
 
-    _autotools = None
+    def export_sources(self):
+        export_conandata_patches(self)
+
+    def validate(self):
```
Comment (SpaceIm) [inline recipes/cunit/all/conanfile.py:44]: validate() after requirements() please

File: recipes/cunit/all/conanfile.py [hunk 12]
```diff
@@ -30,116 +37,113 @@ class CunitConan(ConanFile):
...
-        with self._build_context():
-            with tools.chdir(self._source_subfolder):
-                autotools = self._configure_autotools()
-                autotools.install()
+        copy(self, "COPYING", self.source_folder, os.path.join(self.package_folder, "licenses"))
+
+        autotools = Autotools(self)
+        autotools.install()
 
         if self.settings.compiler == "Visual Studio" and self.options.shared:
```
Comment (SpaceIm) [inline recipes/cunit/all/conanfile.py:139]:
```suggestion
        if is_msvc(self) and self.options.shared:
```

**Comment by Pro** [comment]
@franramirez688 @SpaceIm ping. You were assigned for review by the bot

**Comment by samuel-emrys** [comment]
@Pro thanks for your work on this. Could you update `test_package/conanfile.py` to resolve the conan 2 errors while you're at it?

**Review by samuel-emrys — CHANGES_REQUESTED** [review]
@Pro thanks for your work on this. Could you update `test_package/conanfile.py` to resolve the conan 2 errors while you're at it?

File: recipes/cunit/all/conandata.yml [hunk 13]
```diff
@@ -5,6 +5,11 @@ sources:
...
     - patch_file: "patches/0001-configure-in-msvc-shortcuts.patch"
-      base_path: "source_subfolder"
+      patch_description: "configure snprintf for msvc"
+      patch_type: "conan"
     - patch_file: "patches/0002-snprintf-msvc.patch"
-      base_path: "source_subfolder"
+      patch_description: "only set snprintf for older VS compiler"
+      patch_type: "conan"
+    - patch_file: "patches/0003-fix-copy-docs.patch"
+      patch_description: "use correct docs dir for install docs"
```
Comment (memsharded) [inline recipes/cunit/all/conandata.yml:14]: But installing docs shouldn't be a thing, is it? Docs are not installed or packaged in ConanCenter packages.

**Review by memsharded — COMMENTED** [review]
Thanks very much for your contribution, and sorry for the delays in reviewing. The backlog is very large, and it is challenging to manage everything.

I am assigning this to myself to try to move it forward, and I will also try to make it Conan 2.0 compatible.

**Comment by memsharded** [comment]
I have fixed the errors in ``test_package``, now it is green for both 1.X and 2.0.

Could you please sign again the CLA @Pro? It was reset and now it needs to be signed again, sorry for the inconvenience.

**Review by memsharded — APPROVED** [review]

**Review by samuel-emrys — APPROVED** [review]

**Comment by ThijsSassen** [comment]
@memsharded with these changes i now get an error when installing cunit 2.1.3 when using conan 1.59

cunit/2.1-3:INVALID - Invalid
ERROR: There are invalid packages (packages that cannot exist for this configuration):
cunit/2.1-3: Invalid ID: cunit package is built with shared, which requires cunit:with_curses=ncurses option

This is on Ubuntu 20.04 64 bit with gcc 9 using Release build type.

**Comment by samuel-emrys** [comment]
> @memsharded with these changes i now get an error when installing cunit 2.1.3 when using conan 1.59
> 
> cunit/2.1-3:INVALID - Invalid ERROR: There are invalid packages (packages that cannot exist for this configuration): cunit/2.1-3: Invalid ID: cunit package is built with shared, which requires cunit:with_curses=ncurses option
> 
> This is on Ubuntu 20.04 64 bit with gcc 9 using Release build type.

You need to follow the directions of the error message:

```
conan install --requires=cunit/2.1-3 -o "cunit/*:shared=True" -o "cunit/*:with_curses=ncurses"
```

**Pull request merged**